sys/include/net/gcoap/dns.h

```diff
@@ -0,0 +1,200 @@
+ *                          address will take preference over an IPv4 address).
+ *
+ * @return  length of @p addr_out in bytes on success
+ * @return  -EAFNOSUPPORT, if the hostname of the URI resolves to an unknown address family.
+ * @return  -EBADMSG, when receiving erroneous response or response containing
+ *          an error code.
+ * @return  -ECONNABORTED, if CoAP request cannot be sent.
+ * @return  -ECONNREFUSED, if no URI is set for the client (see @ref gcoap_dns_server_uri_set()).
+ * @return  -EDESTADDRREQ, if CoAP response was received from an unexpected
+ *          remote.
+ * @return  -EHOSTUNREACH, if the hostname of the URI can not be resolved
+ * @return  -ENOBUFS, if there was not enough buffer space for the request.
+ * @return  -ENOBUFS, if length of received CoAP body is greater than
+ *          @ref CONFIG_DNS_MSG_LEN.
+ * @return  -ENOENT, if Zone-ID of the URI can not be found locally.
+ * @return  -ENOMSG, if CoAP response did not contain a DNS response.
+ * @return  -ENOTRECOVERABLE, on gCoAP-internal error.
+ * @return  -ENOTSUP, if credential can not be added for to client.
+ * @return  -ENOTSUP, if module `gcoap_dns_async` is not compiled in.
+ * @return  -ETIMEDOUT, if CoAP request timed out.
+ */
+int gcoap_dns_query(const char *domain_name, void *addr_out, int family);
+
+/**
+ * @brief   Sets and checks a URI
+ *
+ * @param[in] uri   A URI. May be NULL to remove the URI.
+ *
+ * @return  length of @p uri on success (0 when URI is removed).
+ * @return  -EINVAL, if @p uri is not a valid URI for DNS over CoAP.
+ * @return  -ENOBUFS, if the client implementation has not enough buffer space
+ *          to process the URI.
+ */
+int gcoap_dns_server_uri_set(const char *uri);
+
+/**
+ * @brief   Gets the URI
+ *
+ * This function is not thread-safe.
+ *
+ * @return  The current URI for the client.
+ */
+const char *gcoap_dns_server_uri_get(void);
+
+/**
+ * @brief   Deletes all added credentials
+ *
+ * This also removes the credentials from the @ref net_credman and the GCoAP DTLS sock.
+ */
+void gcoap_dns_cred_reset(void);
+
+/**
+ * @brief   Adds a credential for the use with the configured DoC server.
+ *
+ * @pre creds != NULL.
+ *
+ * @param[in] creds     A crential. May not be NULL.
+ *
+ * @return 0 on success
+ * @return -ENOTSUP if DTLS is not supported (because `gcoap_dtls` is not compiled in)
+ * @return -ENOMEM if @ref CONFIG_GCOAP_DNS_CREDS_MAX is exceeded (duplicates may be added).
+ * @return -EBADF if the credential can not be added to @ref net_credman or the GCoAP DTLS sock.
+ */
+int gcoap_dns_cred_add(credman_credential_t *creds);
+
+/**
+ * @brief   Remove a credential for the use with the configured DoC server
+ *
+ * This also removes the credentials from the @ref net_credman and the GCoAP DTLS sock.
+ *
+ * @param[in] tag   The tag of the credential.
+ * @param[in] type  The type of the credential.
+ */
+void gcoap_dns_cred_remove(credman_tag_t tag, credman_type_t type);
+
+/**
```

Comment:
Is this functionality relevant enough that there is a dedicated "set proxy for X" function group for DNS-over-CoAP? Might it not make more sense to have a system-global proxy, or a more general mechanism to pick the right proxy for a request (into which the application that is sending the request may provide input)?

Reply:
True, but at the moment this functionality does not exist, so I would prefer using the current approach and to keep it, in case future experimentation is needed.